File: code/python/123.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        def maxProfitForK(prices,k):
            dp_i_0 = [0]*(k+1)#每一天交易k次的时候，当前未持有股票能获得的最大利润
            dp_i_1 = [-float('inf')]*(k+1)#每一天k次。。。
            for price in prices:
                for i in range(1,k+1):
                    dp_i_0[i] = max(dp_i_0[i],dp_i_1[i]+price)
                    dp_i_1[i] = max(dp_i_1[i],dp_i_0[i-1]-price)
            return dp_i_0[-1]
        return maxProfitForK(prices,2)

File: code/java/123.java
class Solution {
    public int maxProfit(int[] prices) {
        int K = 2;

        int[] dp_0 = new int[K+1];
        int[] dp_1 = new int[K+1];
        for(int i=0;i<=K;i++) dp_1[i] = Integer.MIN_VALUE;
        for(int i=0;i<prices.length;i++){
            for(int k=1;k<=K;k++){
                int temp = dp_0[k];
                dp_0[k] = Math.max(dp_0[k],dp_1[k]+prices[i]);
                dp_1[k] = Math.max(dp_1[k],dp_0[k-1]-prices[i]);
            }
        }
        return dp_0[K];
    }
}